crates/runtime/src/instance/allocator/pooling/index_allocator.rs

```diff
@@ -0,0 +1,408 @@
+//! Index/slot allocator policies for the pooling allocator.
+
+use super::PoolingAllocationStrategy;
+use crate::CompiledModuleId;
+use rand::Rng;
+use std::collections::HashMap;
+
+#[derive(Clone, Debug)]
+pub(crate) enum PoolingAllocationState {
+    NextAvailable(Vec<usize>),
+    Random(Vec<usize>),
+    /// Reuse-affinity policy state.
+    ///
+    /// The data structures here deserve a little explanation:
+    ///
+    /// - free_list: this is a vec of slot indices that are free, no
+    ///   matter their affinities.
+    /// - per_module: this is a hashmap of vecs of slot indices that
+    ///   are free, with affinity for particular module IDs. A slot may
+    ///   appear in zero or one of these lists.
+    /// - slot_state: indicates what state each slot is in: allocated
+    ///   (Taken), only in free_list (Empty), or in free_list and a
+    ///   per_module list (Affinity).
+    ///
+    /// The slot state tracks a slot's index in the global and
+    /// per-module freelists, so it can be efficiently removed from
+    /// both. We take some care to keep these up-to-date as well.
+    ///
+    /// On allocation, we first try to find a slot with affinity for
+    /// the given module ID, if any. If not, we pick a random slot
+    /// ID. This random choice is unbiased across all free slots.
+    ReuseAffinity {
+        // Free-list of all slots. We use this to pick a victim when
+        // we don't have an appropriate slot with the preferred
+        // affinity.
+        free_list: Vec<usize>,
+        // Invariant: any module ID in this hashmap must have a
+        // non-empty list of free slots (otherwise we remove it).
+        per_module: HashMap<CompiledModuleId, Vec<usize>>,
```

Comment:
Is the idea that as `Module`s get dropped, we don't need to actively empty the `per_module` free lists, because they will eventually get stolen by other instance allocations, and once the dropped module's free list is empty, it is removed from the `per_module` free lists? I think this would be good to test somehow, since it seems like an easy memory leak if we mess this up. Also, it seems like it would be good to document in a comment somewhere.

Reply:
Yep, exactly; a slot's affinity remains until the slot is overwritten (reused for a different module). Extended the doc-comment, and added some testing for the keys of this hashmap (and that its values are non-empty lists) as well.